Подключим необходимые библиотеки

In [1]:
library(Biobase)
library(GEOquery)
library(biomaRt)

Loading required package: BiocGenerics
Loading required package: parallel

Attaching package: 'BiocGenerics'

The following objects are masked from 'package:parallel':

    clusterApply, clusterApplyLB, clusterCall, clusterEvalQ,
    clusterExport, clusterMap, parApply, parCapply, parLapply,
    parLapplyLB, parRapply, parSapply, parSapplyLB

The following objects are masked from 'package:stats':

    IQR, mad, sd, var, xtabs

The following objects are masked from 'package:base':

    anyDuplicated, append, as.data.frame, basename, cbind, colnames,
    dirname, do.call, duplicated, eval, evalq, Filter, Find, get, grep,
    grepl, intersect, is.unsorted, lapply, Map, mapply, match, mget,
    order, paste, pmax, pmax.int, pmin, pmin.int, Position, rank,
    rbind, Reduce, rownames, sapply, setdiff, sort, table, tapply,
    union, unique, unsplit, which, which.max, which.min

Welcome to Bioconductor

    Vignettes contain introductory material; view with
    'browseVignettes()'. To cite B

Считаем данные эскперимента

In [2]:
gset <- getGEO("GSE26728", GSEMatrix =TRUE, getGPL=FALSE)
if (length(gset) > 1) idx <- grep("GPL7042", attr(gset, "names")) else idx <- 1
gset <- gset[[idx]]

Found 1 file(s)
GSE26728_series_matrix.txt.gz
Parsed with column specification:
cols(
  ID_REF = col_character(),
  GSM658075 = col_double(),
  GSM658076 = col_double(),
  GSM658077 = col_double(),
  GSM658078 = col_double(),
  GSM658079 = col_double(),
  GSM658080 = col_double(),
  GSM658081 = col_double(),
  GSM658082 = col_double(),
  GSM658083 = col_double(),
  GSM658084 = col_double(),
  GSM658085 = col_double(),
  GSM658086 = col_double(),
  GSM658087 = col_double(),
  GSM658088 = col_double(),
  GSM658089 = col_double(),
  GSM658090 = col_double(),
  GSM658091 = col_double(),
  GSM658092 = col_double()
)


Посмотрим на матрицу экспрессии генов. Здесь названия строк - идентификаторы пробы, названия столбцов - особи.

In [3]:
ex_mat <- exprs(gset)

In [4]:
head(ex_mat)

GSM658075 GSM658076 GSM658077 GSM658078 GSM658079 GSM658080
A_51_P100034 12.630    11.899    12.393    12.539    12.368    12.361   
A_51_P100063  5.828     6.598     5.984     5.986     6.089     6.163   
A_51_P100099  7.638     7.635     7.216     7.767     7.516     7.378   
A_51_P100155 10.947    10.582    10.406    11.054    10.770    10.402   
A_51_P100174  6.632     6.089     6.012     6.190     6.260     6.131   
A_51_P100181  6.655     6.788     6.584     6.822     6.742     6.782   
             GSM658081 GSM658082 GSM658083 GSM658084 GSM658085 GSM658086
A_51_P100034 12.332    12.148    12.032    11.955    12.222    12.202   
A_51_P100063  6.384     6.410     5.984     6.158     6.031     5.833   
A_51_P100099  7.171     7.558     7.571     7.607     7.523     7.628   
A_51_P100155 10.610    10.512    10.492    10.485    10.839    10.535   
A_51_P100174  5.993     6.274     6.002     6.011     6.124     6.062   
A_51_P100181  6.805     6.822     6.734     6.954     6.777     6.859   
             GSM658087 GSM658088 GSM658089 GSM658090 GSM658091 GSM658092
A_51_P100034 12.107    12.529    11.838    12.054    12.223    12.343   
A_51_P100063  6.441     7.814     6.107     7.017     5.853     6.656   
A_51_P100099  7.461     7.321     7.792     7.577     7.412     7.357   
A_51_P100155 10.674    10.333    10.622    10.302    10.339    10.572   
A_51_P100174  6.375     6.087     6.152     6.202     6.209     6.172   
A_51_P100181  6.684     6.600     6.779     7.029     6.616     6.715

In [5]:
ag_probs <- rownames(ex_mat)
samples <- colnames(ex_mat)

Проаннотируем пробы, чтобы получить информацию об исследуемых генах. Для этого воспользуемся пакетом BiomaRT.

In [8]:
mart <- useDataset("mmusculus_gene_ensembl", useMart("ensembl"))
gene_list <- getBM(attributes= c("agilent_wholegenome", "ensembl_gene_id"), values=ag_probs, filter = "agilent_wholegenome", mart= mart)

Cache found


In [9]:
head(gene_list)

agilent_wholegenome ensembl_gene_id   
1 A_51_P101086        ENSMUSG00000024276
2 A_51_P122981        ENSMUSG00000104526
3 A_51_P119633        ENSMUSG00000067199
4 A_51_P129480        ENSMUSG00000095675
5 A_51_P133792        ENSMUSG00000031145
6 A_51_P105755        ENSMUSG00000012640

Посмотрим все ли пробы получилось проаннотировать. Сравним длины исходных и полученных данных:

In [10]:
number_of_not_converted <- length(ag_probs) - length(gene_list$agilent_wholegenome)
number_of_not_converted

[1] 1761

Для 1761 элемента не нашлось соответствий в базе данных. 
Дальнейшие действия будем выполнять без их учета.

Известно, что первые 6 особей соответствуют группе "control", следующие - "BPA at 50" и последние 6 - "BPA at 5000". Найдем средние значения экспрессии генов для каждой из групп.

In [11]:
means_for_genes = data.frame(gene=unique(gene_list$ensembl_gene_id))
means_for_genes$control <- rep(0, nrow(means_for_genes))
means_for_genes$bpa_at_50 <- rep(0, nrow(means_for_genes))
means_for_genes$bpa_at_5000 <- rep(0, nrow(means_for_genes))

In [12]:
for (i in 1:nrow(gene_list)){
    probs_for_gene <- as.vector(gene_list[gene_list$ensembl_gene_id == gene_list$ensembl_gene_id[i],]$agilent_wholegenome)
    means_for_genes[means_for_genes$gene == gene_list$ensembl_gene_id[i],]$control = mean(ex_mat[probs_for_gene, c(1:6)]) / length(probs_for_gene)
    means_for_genes[means_for_genes$gene == gene_list$ensembl_gene_id[i],]$bpa_at_50 = mean(ex_mat[probs_for_gene, c(7:12)]) / length(probs_for_gene)
    means_for_genes[means_for_genes$gene == gene_list$ensembl_gene_id[i],]$bpa_at_5000 = mean(ex_mat[probs_for_gene, c(13:18)]) / length(probs_for_gene)
}

In [13]:
means_for_genes

gene               control   bpa_at_50 bpa_at_5000
1     ENSMUSG00000024276  3.503208  3.543042  3.496792  
2     ENSMUSG00000104526  6.938000  6.886667  7.115833  
3     ENSMUSG00000067199  4.430417  4.150042  4.361042  
4     ENSMUSG00000095675  6.663333  6.606833  6.965167  
5     ENSMUSG00000031145  7.433667  7.355167  7.270333  
6     ENSMUSG00000012640  2.208241  2.200611  2.205481  
7     ENSMUSG00000028062 13.264333 13.302667 13.122167  
8     ENSMUSG00000068566  3.322625  3.390000  3.285542  
9     ENSMUSG00000039826  4.968250  5.032208  5.050667  
10    ENSMUSG00000051147 10.281833 10.246167 10.291500  
11    ENSMUSG00000028407  3.726958  3.749833  3.771000  
12    ENSMUSG00000026816  9.456167  9.606500  9.554833  
13    ENSMUSG00000028156  3.002741  3.054722  3.016093  
14    ENSMUSG00000110528  7.706500  7.908833  7.748667  
15    ENSMUSG00000041556  2.659907  2.763685  2.747759  
16    ENSMUSG00000042197  1.620552  1.610010  1.620385  
17    ENSMUSG00000038319  6.183500  6.353000  6.235333  
18    ENSMUSG00000047037  8.304000  8.497833  8.506667  
19    ENSMUSG00000027677  2.640389  2.648574  2.656556  
20    ENSMUSG00000096596  6.663333  6.606833  6.965167  
21    ENSMUSG00000078515  6.135917  5.999292  6.147042  
22    ENSMUSG00000030423  4.643583  4.777417  4.741042  
23    ENSMUSG00000059934  9.302000  9.333667  9.362333  
24    ENSMUSG00000031889  9.319667  9.456000  9.401667  
25    ENSMUSG00000034211  5.085708  5.074542  5.086958  
26    ENSMUSG00000101188  3.534352  3.520741  3.543481  
27    ENSMUSG00000069266  5.148167  5.277250  5.196625  
28    ENSMUSG00000060093  3.278593  3.354593  3.301963  
29    ENSMUSG00000114078  6.355500  6.247667  6.227000  
30    ENSMUSG00000056290  3.145875  3.177708  3.164417  
...   ...                ...       ...       ...        
14042 ENSMUSG00000024163  6.048333  6.209833  5.945000  
14043 ENSMUSG00000024245  6.153833  6.142500  6.121667  
14044 ENSMUSG00000094520  6.560500  6.450667  6.478333  
14045 ENSMUSG00000074577  5.720333  5.760000  5.806833  
14046 ENSMUSG00000030695 13.855500 13.625667 13.552333  
14047 ENSMUSG00000047841  6.071333  6.038667  6.165167  
14048 ENSMUSG00000080962 13.562833 13.745167 13.521500  
14049 ENSMUSG00000001576  4.601417  4.682708  4.597417  
14050 ENSMUSG00000074570  5.781000  5.815667  5.850000  
14051 ENSMUSG00000021221  5.863500  6.007000  5.973333  
14052 ENSMUSG00000094910  7.119667  7.176667  7.087000  
14053 ENSMUSG00000079429  6.230333  6.196833  6.355667  
14054 ENSMUSG00000103763  6.035500  5.973333  6.098833  
14055 ENSMUSG00000073374  6.461000  6.606333  6.645667  
14056 ENSMUSG00000056014  5.815333  5.685500  5.765167  
14057 ENSMUSG00000027111  6.076667  6.168000  6.011500  
14058 ENSMUSG00000025357  8.547500  8.310500  8.557667  
14059 ENSMUSG00000045287  8.513167  8.162500  8.323500  
14060 ENSMUSG00000025354  6.028167  6.069500  6.043667  
14061 ENSMUSG00000025347 12.101667 12.448833 11.875333  
14062 ENSMUSG00000098339  5.798500  5.751833  5.738667  
14063 ENSMUSG00000065738  6.348833  6.227167  6.092500  
14064 ENSMUSG00000064829  7.154333  6.965167  6.952167  
14065 ENSMUSG00000093149  9.731500  9.960500  9.914833  
14066 ENSMUSG00000064731  6.348833  6.227167  6.092500  
14067 ENSMUSG00000065295  7.154333  6.965167  6.952167  
14068 ENSMUSG00000046747  5.876333  5.913500  5.760167  
14069 ENSMUSG00000116683  6.444167  5.983000  6.196167  
14070 ENSMUSG00000107119  8.033833  8.034333  7.945667  
14071 ENSMUSG00000111592 14.223167 13.914333 13.965667

Сохраним результат в файл

In [14]:
write.table(means_for_genes, "../data/means_for_genes.txt", row.names = FALSE, col.names = TRUE, sep='\t', quote=FALSE)

Так же запишем в файл матрицу экспрессии для использования в дальнейшем для построения дерева решений.

In [15]:
df = as.data.frame(t(ex_mat))
df$Group <- c(rep(0, 6), rep(1, 6), rep(2, 6))

In [16]:
write.table(df, "../data/expr_matrix.txt", row.names = TRUE, col.names = TRUE, sep='\t', quote=FALSE)